mm2src/mm2_net/src/wasm/body_stream.rs

```diff
@@ -0,0 +1,398 @@
+                    }
+                },
+                ReadState::TrailerLength => {
+                    if this.buf.len() < 4 {
+                        // Can't read data length right now
+                        return Ok(());
+                    } else {
+                        let trailer_length_bytes = this.buf.take(4);
+                        let trailer_length = BigEndian::read_u32(trailer_length_bytes.as_ref());
+                        *this.state = ReadState::Trailer(trailer_length);
+                    }
+                },
+                ReadState::Trailer(trailer_length) => {
+                    let trailer_length = *trailer_length as usize;
+
+                    if this.buf.len() < trailer_length {
+                        // Can't read trailer right now
+                        return Ok(());
+                    } else {
+                        let mut trailer_bytes = this.buf.take(trailer_length);
+                        trailer_bytes.put_u8(b'\n');
+
+                        let mut trailers_buf = [EMPTY_HEADER; 64];
+                        let parsed_trailers = match httparse::parse_headers(&trailer_bytes, &mut trailers_buf)
+                            .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?
+                        {
+                            Status::Complete((_, headers)) => Ok(headers),
+                            Status::Partial => Err(PostGrpcWebErr::InvalidRequest("Invalid".to_string())),
+                        }?;
+
+                        let mut trailers = HeaderMap::with_capacity(parsed_trailers.len());
+
+                        for parsed_trailer in parsed_trailers {
+                            let header_name = HeaderName::from_bytes(parsed_trailer.name.as_bytes())
+                                .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?;
+                            let header_value = HeaderValue::from_bytes(parsed_trailer.value)
+                                .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?;
+                            trailers.insert(header_name, header_value);
+                        }
+
+                        *this.trailer = Some(trailers);
+
+                        *this.state = ReadState::Done;
+                    }
+                },
+                ReadState::Done => return Ok(()),
+            }
+        }
+    }
+}
+
+impl Body for ResponseBody {
+    type Data = Bytes;
+
+    type Error = PostGrpcWebErr;
+
+    fn poll_data(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        // Check if there's already some data in buffer and return that
+        if self.data.is_some() {
+            return Poll::Ready(self.data.take().map(|d| Ok(d.freeze())));
+        }
+
+        // If reading data is finished return `None`
+        if self.state.finished_data() {
+            return Poll::Ready(None);
+        }
```

Comment:
The proposal was to replace *both* `if self.data.is_some()` and `if self.state.finished_data()` with
```rust
// If reading data is finished return it
if self.state.finished_data() {
    return Poll::Ready(self.data.take().map(|data| Ok(data.freeze())));
}
```

If we finished reading data, the replacement code covers both `Some` and `None` cases.